src/CodeGen_C.cpp

```diff
@@ -181,7 +179,38 @@ const string preamble =
     "}\n";
 }
 
-CodeGen_C::CodeGen_C(ostream &s) : IRPrinter(s), id("$$ BAD ID $$") {}
+CodeGen_C::CodeGen_C(ostream &s, bool is_header, const std::string &guard) : IRPrinter(s), id("$$ BAD ID $$"), is_header(is_header) {
+    if (is_header) {
+        // If it's a header, emit an include guard.
+        stream << "#ifndef HALIDE_" << print_name(guard) << '\n'
+               << "#define HALIDE_" << print_name(guard) << '\n';
+    }
+
+    stream << "#ifdef __cplusplus\n";
+    stream << "#define HALIDE_EXTERN_C extern \"C\"\n";
+    stream << "#else\n";
+    stream << "#define HALIDE_EXTERN_C\n";
```

Comment:
Why are we using a macro at all instead of opening an extern C block before the definitions and closing it afterwards and protecting the block open and close with the standard C++ #ifdef conditional?